### ISSUE TITLE
Phase11-Part1-MohammadReza-Sadeghi

### ISSUE THREAD

**Issue by MRSadeghi78** [Star-Academy/codestar-intern-issues#191]
This issue template is based of [Phase11-Web.md](./Phase11-Web.md) contents.
Click on "Raw" button on right side of this page to see original markdown then create a new issue with original content of this tasklist:

# Phase 11 First Part

- [x] Before starting: Read the link about history and purpose of creating & developing Web, then answer this questions:
  - [x] How did developers provide server-client softwares for users before creation of web technologies?
  
      `Each client-server program must be compiled and installed in a specific OS so they were completely dependent to the operating systems`
      
  - [x] What was the main trouble with traditional technologies before web?
  
      `The first problem was that Both client and server apps were too dependent on the user's OS so it was necessary to write a distinct program for each OS. The second problem was that server had to send all the needed data to the client side so that the client program was able to display the content for the user, which was expensive and slow.`
      
  
  - [x] How did web overcome this main trouble?
  
      `The first problem solved by browsers so that the only part that was dependent on the OS was the browser and other parts are completely independent of OS. The second problem solved by javascript so that some part of web pages content was displayed by javascript.`

## Asp.net core

- [x] Getting Started with HTTP:
  - [x] What does a protocol mean?
  
    `A set of rules used in programming computers so that they can communicate with each other`

  - [x] Why designing a protocol for communication is important prerequirement of all other web related technologies?
  
    `Because two computers or programs can not communicate with each other if they don't know the other one will send data in which format`
    
  - [x] Try to get familier with HTTP by reading links.
  - [x] Why HTTP is most popular protocol in web technologies?
    
    `1. It's simple`
    `2. It's text-format. This makes it easier for us to test it`
    `3. It is extensible so it can transfer multi media contents too.`

  - [x] What is the most interesting aspect of HTTP in your opinion? 
  
    `It's human readable so we can learn and test it simply.`

- [x] Getting Started with REST:
  - [x] Read mentioned links.
  - [x] Why do we need RESTful APIs?
  
    `1. Due to it's scalability`
    `2. Due to its flexibility and portability.`
    `3. Due to its independence.`

- [x] Getting Started with Asp.net Core:
  - [x] Read mentioned links.
  - [x] Create a hello world API.
  - [x] Call your api from browser.
  - [x] Play with routing options! (Try to use it & enjoy it!)
  - [x] Why do we need configurable routings in web development?
  
    `Because we want each request to be mapped to a certain action and we achieve this by using routings.`
  
  - [x] Play with dependency injection! (Define an interface & inject interface implementation to code)
  - [x] Why do we need dependency injection?
  
    `Because we want to minimize the dependency in different parts of our code. Therefore we use interfaces so that if we change a part of the class, other parts of the code remain unchanged.`

  - [x] Play with parameter binding! (use different modes)
  - [x] Install postman.
  - [x] Call your APIs with postman.
  
- [x] Implement Asp.net Core Project:
  - [x] Design your search API.
  - [x] Implement & Test your API.

# Review
Link of your own PR:

[My PR](https://github.com/Star-Academy/Winter99-Codestar-MohammadReza-Sadeghi/pull/9)

Link of other PR(s) you agreed to review:

`[FILL HERE WITH LINK OF PR(s) YOU REVIEWED.]`

- [ ] Your own PR is reviewed and approved by at least one other team.

- [ ] Your own PR is reviewed and approved by mentor (hamyar).

- [ ] Your own PR is reviewed and approved by manager (rahbar).

- [ ] Your own PR is merged.
